third_party/nvfuser/csrc/scheduler/normalization.cpp

```diff
@@ -21,15 +21,178 @@ namespace nvfuser {
+std::shared_ptr<ReductionParams> innerOuterPersistentHeuristic(
+    const int64_t total_reduction_numel,
+    const int64_t total_iteration_numel,
+    const int64_t inner_most_dimension_numel,
+    const int64_t n_tensor_inputs,
+    const int64_t max_input_dtype_size,
+    const int64_t max_persistent_buffer_size,
+    const size_t vectorize_factor) {
+  auto rparams = std::make_shared<ReductionParams>();
+  rparams->persistent_kernel = true;
+  rparams->fastest_dim = true;
+  rparams->combined_inner_outer = true;
+  if (inner_most_dimension_numel <= 1024) {
+    rparams->multiple_reds_per_blk = true;
+  }
+
+  const auto dev_prop = at::cuda::getCurrentDeviceProperties();
+  const int64_t device_multiprocessor_count =
+      (int64_t)dev_prop->multiProcessorCount;
+
+  auto getRegPerThread = [&](const int threadsPerBlock, const int blocksPerSM) {
+    int64_t warpsPerSM =
+        ceilDiv(threadsPerBlock, dev_prop->warpSize) * blocksPerSM;
+    constexpr int reg_allocation_granularity = 256;
+    int64_t regPerWarp = dev_prop->regsPerMultiprocessor / warpsPerSM /
+        reg_allocation_granularity * reg_allocation_granularity;
+    return regPerWarp / dev_prop->warpSize;
+  };
+
+  int64_t vect = vectorize_factor;
+  int64_t after_vect = ceilDiv(inner_most_dimension_numel, vect);
+  // split input_vect between threadsPerBlock and batch
+  int64_t threadsPerBlock = 32;
+  int64_t batch = ceilDiv(after_vect, threadsPerBlock);
+  int64_t batchMax = 20;
+  int64_t batchMin;
+  if (inner_most_dimension_numel >= 3072) {
+    // allow small batch for small iter numbers, e.g. shape= 2048 x 3072
+    if (total_iteration_numel <= 2048 && inner_most_dimension_numel == 3072) {
+      batchMin = 3;
+    } else {
+      batchMin = 4;
+    }
+  } else if (inner_most_dimension_numel >= 2048) {
+    batchMin = 2;
+  } else {
+    batchMin = 1;
+  }
+  int64_t threadsPerBlockMax = inner_most_dimension_numel >= 20480 ? 512 : 256;
+  while (batch > batchMin && threadsPerBlock < threadsPerBlockMax) {
+    if (batch % 2 == 0 && batch / 2 >= batchMin) {
+      batch /= 2;
+      threadsPerBlock *= 2;
+    } else if (batch % 3 == 0 && batch / 3 >= batchMin) {
+      batch /= 3;
+      threadsPerBlock *= 3;
+    } else if (batch % 5 == 0 && batch / 5 >= batchMin) {
+      batch /= 5;
+      threadsPerBlock *= 5;
+    } else if (batch > batchMax) {
+      threadsPerBlock = scheduler_utils::roundUpToN(
+          ceilDiv(after_vect, batchMax), dev_prop->warpSize);
+      batch = ceilDiv(after_vect, threadsPerBlock);
+      break;
+    } else {
+      break;
+    }
+  }
+  // set gdimy & bdimx w.c.  gdimy*unroll*bdimx >= inner_most_dimension_numel
+  int64_t unroll_outer_reduction = inner_most_dimension_numel >= 4096 ? 4 : 2;
+  int64_t after_unroll =
+      ceilDiv(inner_most_dimension_numel, unroll_outer_reduction);
+  int64_t blocksPerSmMax1 = ceilDiv(after_unroll, device_multiprocessor_count);
+
+  int64_t regPerThread = std::min(batch * vect * 8, (int64_t)255);
```

Comment:
Why multiplication by 8?

Reply:
This was based on my observation of PTX info. Revised estimation using persistent buffers.